includes/class-amp-story-post-type.php

```diff
@@ -1490,6 +1522,232 @@ public static function insert_attachment( $object, $cropped ) {
+
+			// Passed to `get_preview_post_link()` for nonce access and to sanitize the output.
+			$query_args = array(
+				'story_export' => $slug,
+				'_wpnonce'     => wp_create_nonce( 'amp-story-export-' . $post->ID ),
+			);
+
+			// Passed to `wp_remote_get()`.
+			$args = array(
+				'timeout'   => 30,
+				'sslverify' => false,
+			);
+
+			// Get the preview URL.
+			$response = wp_remote_get( get_preview_post_link( $post, $query_args ), $args );
+
+			// Ensure we have the required data.
+			if ( ! ( is_array( $response ) && isset( $response['body'] ) && isset( $response['headers']['x-amp-export-assets'] ) ) ) {
+				return new WP_Error( 'amp_story_export_response', esc_html__( 'Could not get the AMP Story HTML.', 'amp' ) );
+			}
+
+			// Add index.html file.
+			$zip->addFromString( $slug . '/index.html', $response['body'] );
+
+			// Get the assets.
+			$assets = explode( ',', $response['headers']['x-amp-export-assets'] );
+
+			// Add the assets.
+			if ( ! empty( $assets ) ) {
+
+				// Create the empty assets directory.
+				$zip->addEmptyDir( $slug . '/assets' );
+
+				foreach ( $assets as $asset ) {
+					$contents = file_get_contents( $asset ); // phpcs:ignore WordPress.WP.AlternativeFunctions.file_get_contents_file_get_contents
+					if ( false !== $contents ) {
+						$zip->addFromString( $slug . '/assets/' . basename( $asset ), $contents );
+					}
+				}
+			}
+
+			// Close the active archive.
+			$zip->close();
+
+			// Read the file.
+			$fo = @fopen( $file, 'r' ); // phpcs:ignore
+
+			if ( ! $fo ) {
+				return new WP_Error( 'amp_story_export_file_open', esc_html__( 'Could not open the generated ZIP archive.', 'amp' ) );
+			} else {
+				header( 'Content-type: application/octet-stream' );
+				header( 'Content-Disposition: attachment; filename="' . $file . '"' );
+				header( 'Content-length: ' . filesize( $file ) );
+				fpassthru( $fo );
+				unlink( $file );
+				die();
+			}
+		} else {
+			switch ( $res ) {
+				case ZipArchive::ER_EXISTS:
+					$msg = esc_html__( 'Zip archive already exists.', 'amp' );
+					break;
+				case ZipArchive::ER_INCONS:
+					$msg = esc_html__( 'Zip archive inconsistent.', 'amp' );
+					break;
+				case ZipArchive::ER_MEMORY:
+					$msg = esc_html__( 'Malloc failure.', 'amp' );
+					break;
+				case ZipArchive::ER_NOENT:
+					$msg = esc_html__( 'No such zip archive.', 'amp' );
+					break;
+				case ZipArchive::ER_NOZIP:
+					$msg = esc_html__( 'Not a zip archive.', 'amp' );
+					break;
+				case ZipArchive::ER_OPEN:
+					$msg = esc_html__( 'Could not open zip archive.', 'amp' );
+					break;
+				case ZipArchive::ER_READ:
+					$msg = esc_html__( 'Zip archive read error.', 'amp' );
+					break;
+				case ZipArchive::ER_SEEK:
+					$msg = esc_html__( 'Zip archive seek error.', 'amp' );
+					break;
+				default:
+					$msg = esc_html__( 'Unknown zip archive error', 'amp' );
+					break;
+			}
+			return new WP_Error( 'amp_story_zip_archive_error', $msg );
```

Comment:
Given that these error messages are not really actionable for the user, I'd rather do something like this:

```php
/* translators: %s: error code */
$msg = sprintf( __( 'There was an error generating the ZIP archive. Error code: %s', 'amp' ), $res );
```

Good side effect of that: polyglots don't have to translate 10 new strings.